// @framer/parkerhenderson.framer-x-chartjs/code/BarGraph.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";
import { Bar } from "react-chartjs-2";

// Define type of property
interface Props {
  data: object;
  width: number;
  height: number;
  options: object;

  dataLabels: string | string[];

  datasets: number[];
  datasetLabel: string;
  datasetBgColor: string;
  datasetBorderColor: string;
  datasetBorderWidth: number;
  datasetHoverBgColor: string;
  datasetHoverBorderColor: string;
  datasetData: string[] | string;
  mySplit(): any[];
}

export class BarGraph extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    dataLabels: 'January, February, March, April, May, June, July',
    datasetLabel: 'Sample Bar Graph',
    datasetBgColor: '#ff3300',
    datasetBorderColor: '#212121',
    datasetBorderWidth: 2,
    datasetHoverBgColor: '#34b1a3',
    datasetHoverBorderColor: '#7ab112',
    datasetData: '65, 59, 80, 81, 56, 55, 40'
  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    datasetLabel: { type: ControlType.String, title: 'Chart Title'},
    dataLabels: { type: ControlType.String, title: 'Labels'},
    datasetBgColor: {type: ControlType.String, title: 'Data Background Color'},
    datasetBorderColor: {type: ControlType.String, title: 'Data Border Color'},
    datasetBorderWidth: {type: ControlType.Number, title: 'Data Border Width'},
    datasetHoverBgColor: {type: ControlType.String, title: 'Hover Background Color'},
    datasetHoverBorderColor: {type: ControlType.String, title: 'Hover Border Color'},
    datasetData: {type: ControlType.String, title: 'Data'},
  };

  mySplit = (str) => str.split(',');

  render() {
    return (
        <Bar
          data={{
            labels: this.mySplit(this.props.dataLabels),
            datasets: [
              {
                label: this.props.datasetLabel,
                backgroundColor: this.props.datasetBgColor,
                borderColor: this.props.datasetBorderColor,
                borderWidth: this.props.datasetBorderWidth,
                hoverBackgroundColor: this.props.datasetHoverBgColor,
                hoverBorderColor: this.props.datasetHoverBorderColor,
                data: this.mySplit(this.props.datasetData)
              }
            ],
          }}
          width={this.props.width}
          height={this.props.height}
          options={{
            responsive: true,
            maintainAspectRatio: false,
            scales : {
                yAxes : [{
                    ticks : {
                        beginAtZero : true
                    }   
                }]
            }
          }}
        />
      );
  }
}
