// @framer/zach.browser/code/Browser.tsx
import * as React from "react"
import { useEffect, useState } from "react"
import { Frame, Stack, addPropertyControls, ControlType } from "framer"
import { Safari, Chrome } from "./canvas"

export function Browser({
    children,
    width,
    height,
    browser,
    url,
    scrolling,
    browserHeight,
}) {
    const headerHeight = browser === "Safari" ? 38 : 79 // offset differs per browser header
    const content = children[0] // get the first child
    const contentWidth = content ? content.props.width : 1200 // get the width of the child
    let contentHeight = content ? content.props.height : browserHeight // get the width of the child
    if (scrolling) contentHeight = browserHeight // set a fixed height if scrolling is on
    contentHeight += headerHeight // add the header to the height

    // Pull the root domain from the URL and capitalize it
    const pageTitle = url
        .substring(0, url.lastIndexOf("."))
        .replace(/^\w/, c => c.toUpperCase())

    return (
        <Stack width={"100%"} height={"100%"} background={false}>
            <Frame width={contentWidth} height={contentHeight}>
                <div
                    style={{
                        zIndex: 10,
                        position: "relative",
                        height: contentHeight,
                        width: contentWidth,
                        marginTop: headerHeight,
                        overflow: "auto",
                    }}
                >
                    {children}
                </div>
                <Safari
                    visible={browser === "Safari"}
                    URL={url}
                    width={"100%"}
                    height={"100%"}
                    left={0}
                    top={0}
                />
                <Chrome
                    visible={browser === "Chrome"}
                    URL={url}
                    width={"100%"}
                    height={"100%"}
                    left={0}
                    top={0}
                    pageTitle={pageTitle}
                />
            </Frame>
        </Stack>
    )
}

Browser.defaultProps = {
    width: 1400,
    height: 900,
}

addPropertyControls(Browser, {
    browser: {
        type: ControlType.SegmentedEnum,
        title: "Browser",
        defaultValue: "Safari",
        options: ["Safari", "Chrome"],
    },
    url: {
        type: ControlType.String,
        title: "URL",
        defaultValue: "framer.com",
    },
    scrolling: {
        type: ControlType.Boolean,
        title: "Scrolling",
        defaultValue: false,
        enabledTitle: "On",
        disabledTitle: "off",
    },
    browserHeight: {
        type: ControlType.Number,
        title: "Browser Height",
        defaultValue: 800,
        min: 400,
        max: 1600,
        hidden(props) {
            return props.scrolling === false
        },
    },
})
